[1단계 - 음식점 목록] - 지니(손진영) 미션 제출합니다.

jinyoung234 opened the pull request:
안녕하세요 샐리 :) 타입스크립트가 오랜만이라 어색했지만 어찌저찌 구현하여 제출 해봅니다!

가감 없는 피드백 부탁드리겠습니다. 🥹🙏

# 배포 주소

https://jinyoung234.github.io/javascript-lunch/dist

# getting start

```terminal
npm start // dev server execute
```

# 컴포넌트 구조

<img width="869" alt="image" src="https://github.com/woowacourse/javascript-lunch/assets/87177577/3220cab2-baf6-42df-b089-bd7acd731cef">

# 아키텍처

<img width="956" alt="image" src="https://github.com/woowacourse/javascript-lunch/assets/87177577/7ff4210c-bf45-4f2d-a56f-bf12cbddcbe7">

- `View`
  - 컴포넌트 들을 의미하며, 각 컴포넌트 들은 DOM에 연결되었을 때 `connectedCallback`이 호출 되어 `render` 함수(html template(string)을 dom에 commit), `setEvent`(eventListener를 등록)를 실행
  - 반대로 DOM에서 제거되면 `disconnectedCallback`이 실행되어 이벤트 핸들러를 제거
  - 대체로 도메인으로 부터 비즈니스 로직의 결과를 받아 렌더링 하는 형태

- `Domain`
  - 비즈니스 로직을 수행하는 layer
  - storage에게 접근 (데이터를 주고 받음)
  - storage로 부터 받은 데이터를 view에게 전달하는 역할

- `Storage`
  - 데이터를 다루는 DB의 역할을 수행하는 layer
  - Domain Layer와 상호 작용 하여 데이터를 주고 받는 역할을 수행

# 적용한 점
## Local Storage 활용

원래 view에서 받을 데이터에 대해 전역 store를 만들어 전역 상태 관리를 해보려고 했지만 높은 러닝 커브와 데이터 규모가 작다고 판단되어 페어와 함께 진행 하기엔 어려움이 있다고 판단하였습니다.

따라서 domain에 localStorage를 연동하는 형태로 미션을 진행했습니다.

**Restaurant**

[Restaurant link](https://github.com/jinyoung234/javascript-lunch/blob/step1/src/domain/Restaurant/Restaurant.ts)

**RestaurantStorage**

[RestaurantStorage link](https://github.com/jinyoung234/javascript-lunch/blob/step1/src/storages/RestaurantStorage.ts)

 `Storage Layer` 에서는 **마치 DB 처럼 데이터를 추가하고, 빼내는 역할만 수행**하며, `Domain Layer`인 Restaurant에선 **DB에 접근해서 데이터를 가져온 후 이를 가공(sort, filter)하는 역할을 수행**하도록 설정했습니다.

 또한 Domain에서 storage를 `DI(의존성 주입)` **하는 형태로 설정하여 확장성 높은 객체가 되도록 설계**하였습니다.

## 타입스크립트 적용

도메인 뿐 아니라 UI 로직 에도 타입스크립트를 사용해보았습니다.

사용할 땐 `any` 타입과 `type assertion`을 최소화하는 방향으로 진행하였습니다. (다만, FormData를 사용할 때 타입 설정에 어려움을 겪어 type assertion를 추가했습니다. 🥲)

또한, `declaration file(index.d.ts)`**를 추가하여 image 파일에 대한 import를 원활히 할 수 있도록 설정**했습니다.

마지막으로 `유틸리티 타입(Exclude, Omit 등)`과 `타입 가드`를 충분히 활용하여 **type-safe 한 애플리케이션 설계**를 하려 했습니다.

## 웹 컴포넌트

`connectedCallback`과 `disconnectedCallback`과 같은 생명 주기 메서드 사용 및 하나의 역할 부여를 위해 `custom-elements`를 사용했습니다.

또한, `BaseComponent`를 만들어 **공통 로직을 관리**했습니다.

**웹 컴포넌트가 DOM의 commit 되는 과정**은 아래와 같이 진행 됩니다. 

1. HTML parser는 HTML을 파싱하다 LunchApp을 마주친다.
2. LunchApp이 DOM 트리에 추가 된다.
3. DOM에 연결된 직후 connectedCallback이 실행된다.
4. LunchApp의 render 메서드와 setEvent가 실행되어 template이 DOM에 추가되고, 이벤트 리스너가 설정된다.(등록 된 리스너가 없다면 등록되지 않는다.)
5. 나머지 컴포넌트도 1-4의 과정을 통해 DOM에 추가되는 형태다.

# 궁금한 점
## 바닐라 JS 기반 컴포넌트를 사용할 때의 `재 사용성`?

> 재 사용할 수 있는 컴포넌트에 대해 고민하고 적용해본다.

lms에 이러한 요구 사항이 포함되어 있었습니다.

하지만 저는 이번 미션을 진행할 때 `재 사용성` 보다는 `하나의 역할`에 초점을 두어 컴포넌트를 설계했습니다.

가장 큰 이유 중 하나는 `고려해야 할 이유를 크게 찾지 못했다.`였습니다.

```tsx
const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ styleClass, ...rest }: InputProps, ref) => {
    return <input ref={ref} className={styleClass} {...rest} />
  },
)

// 사용
<Input
  placeholder="유튜브 URL 주소"
  className={`w-[330px] bg-gray-900 focus:border-green-light 'border-gray-300' text-sm sm:py-2 pb-[9px] border-b focus:outline-none text-gray-100 placeholder:text-gray-600`}
  onChange={handleInput}
/>
```

React의 경우 React DOM의 이벤트 핸들러를 통해 핸들러를 props에 전달하여 실행 하는 형태로 만들 수 있기 때문에 atomic 한 컴포넌트들도 하나의 역할을 가질 수 있다고 생각했습니다.

다만, 바닐라 JS에서 웹 컴포넌트를 사용할 땐 이벤트 핸들러를 props로 전달할 수 있는 방법을 파악하는 과정에서 어려움을 겪게 되었습니다.

그렇기 때문에, atomic 한 컴포넌트 들을 여러 곳에서 사용하며 모두 이벤트 핸들러를 가지고 있다면 하나의 Input 내에는 여러 개의 이벤트 핸들러를 가지고 있어 여러 곳에 모두 결합하는 문제가 발생할 수 있다고 생각했습니다.

또한, `스타일 재 사용`의 관점에서도 컴포넌트 분리를 고려해볼 수 있겠다고 생각했지만 이 또한 유틸리티 클래스를 쓴다면 해결 될 문제라고 생각했습니다.

만약 바닐라 JS 환경에서 `재 사용 관점에서 컴포넌트를 분리한다.`를 고려하면 어떤 이유 들을 생각해볼 수 있을지 여쭤보고 싶습니다. 

나머지는 코멘트에 추가해놓겠습니다!

jinyoung234 commented on src/components/BaseComponent/BaseComponent.type.ts:
```diff
@@ -0,0 +1,14 @@
+}
+
+export type CustomEventListenerDictionary = {
+  [key: string]: Omit<CustomEventListener, "target">;
```
mapped 타입에서 key를 컴포넌트 내 eventListeners 멤버의 key로 잡고 싶었지만, 계속 실패해서 다른 방법이 있는지 궁금합니다 .. 🥲

jinyoung234 replied:
맵드 타입과 인덱스 시그니처 타입을 헷갈렸네요 .. 🥲 주신 링크 참고해서 다시 공부해 보겠습니다. 감사합니다!!

jinyoung234 commented on src/components/RestaurantAddForm/RestaurantAddForm.ts:
```diff
@@ -0,0 +1,173 @@
+        userInputValues[key] = value;
+      }
+
+      return userInputValues as unknown as RestaurantDetail;
```
FormData 타입이 정해져 있어서 userInputValues의 타입을 RestaurantDetail로 부여하기 위해 `as unknown as RestaurantDetail`로 설정했는데 다른 방법이 있는지 궁금합니다 .. 🥲

jinyoung234 replied:
감사합니다 :) 참고해서 더 보완해보겠습니다 !!

liswktjs reviewed (changes requested):
안녕하세요 지니 
이번에 리뷰를 맡게된 샐리라고 합니다 

타입스크립트 경험이 꽤 많으신것 같아요
오랜만이라고 적어주셨는데도 타입을 잘 활용해서 구현을 해주셨더라고요 👍 
로직들도 깔끔하게 잘 구현해주셔서 코멘트 달 부분들이 많이 없었습니다 

코멘트에는 아래와 같이 적어주셨는데요 
```
재 사용할 수 있는 컴포넌트에 대해 고민하고 적용해본다.

lms에 이러한 요구 사항이 포함되어 있었습니다.

하지만 저는 이번 미션을 진행할 때 재 사용성 보다는 하나의 역할에 초점을 두어 컴포넌트를 설계했습니다.

가장 큰 이유 중 하나는 고려해야 할 이유를 크게 찾지 못했다.였습니다. 
....
만약 바닐라 JS 환경에서 재 사용 관점에서 컴포넌트를 분리한다.를 고려하면 어떤 이유 들을 생각해볼 수 있을지 여쭤보고 싶습니다.
```

라고 질문을 해주셨는데 , 
예를 들어 현재 dropdown이 해당 서비스에서도 2곳에서 사용되고 있는데요 
확장성, 재사용성 없이 구현하게 되면 서비스를 확장하게 될 때 dropdown 이 필요하다면 또 만들어야 되게 됩니다 이렇게 매번 컴포넌트를 새롭게 만들게 되면 
개발 기간도 늘어나고 코드량도 많아지면서 서비스를 관리하기 힘들어집니다 
그러면 바닐라 자스로만 구현하는 웹에서는 이 모든 걸 포기해야할 까요? 
이런 점에서는 재사용성이 필요하다고 생각합니다 


그리고 약간은  주객이 전도된 느낌도 드네요 🥲
이번 미션의 요구사항은 "재사용할 수 있는 컴포넌트에 대해 고민하고 적용해본다"인데 
웹 컴포넌트와 기능들을 구현하는데 중점을 두신 것 같아요 
물론 코드적으로는 깔끔하고 잘 구현했다고 생각 합니다 
코드 실력도 좋으시고 경험도 많으신것 같아요 

하지만 만약, 목표를 이룰 수 없다면 과감하게 웹 컴포넌트를 사용하지 않는 것도 필요하다고 생각합니다 그리고 재사용성을 포기해가면서 까지 웹 컴포넌트를 써서 이 미션을 구현했어야 하는지도 사실 잘 납득이 되지는 않습니다 
우테코라는 학습 과정에 있으신 만큼 코치님들이 미션마다 제안 해주시는 요구사항을 충족시킬 수 있는 방향을 우선적으로 고려해서 기능을 구현하면 좋을 것 같습니다 

글로 쓰다 보니 딱딱하게 느껴지네요 😅
그래도 지니의 코드에서 리뷰해 줄 부분이 재사용성이라고 생각해서 길게 써보았습니다 

추가로 보안하면 좋을 것 같은 기능들 코멘트에 추가하였습니다! 
한 번 확인 부탁드려요~

jinyoung234 commented on src/components/LunchApp/LunchApp.ts:
```diff
@@ -0,0 +1,21 @@
+
+class LunchApp extends BaseComponent {
+  protected render() {
+    this.innerHTML = `
+          <global-navigation-bar></global-navigation-bar>
+          <main>
+            <section class="restaurant-filter-container">
+              <category-dropdown></category-dropdown>            
+              <sort-dropdown></sort-dropdown>
+            </section>
+            <restaurant-list></restaurant-list>
+          </main>
+          <restaurant-add-modal class="modal"></restaurant-add-modal>
+        `;
```
잘 동작할거 같습니다 !! 컴포넌트 내에서 render 메서드의 일관성을 보장하고자 다음과 같이 코드를 구성했던거 같습니다.

다만 샐리가 코멘트 주신 것과 같이 `innerHTML`은 처음 동작할 때 요소를 없애고 다시 추가하는 만큼 `createElement` 방식이 적절하다고 느껴지는거 같습니다.

샐리가 코멘트 주셨던 이유도 혹시 `innerHTML`의 동작과 관련된 의도에서 주셨을지 여쭤봐도 될까요?! 🙇‍♀️

liswktjs replied:
지니의 코드 실력이라면 innerHTML이 xss공격에 취약하다는 사실을 이미 알 고 있을 것 같아서 
여기에서 왜 innerHTML을 썼지? 
혹시 웹 컴포넌트 개발하면서 어떤 이슈를 만났나 의아해서 질문 드렸습니다

jinyoung234 commented on src/components/GlobalNavigationBar/GlobalNavigationBar.ts:
```diff
@@ -0,0 +1,47 @@
+    });
+  }
+
+  private handleOpenModal() {
+    $(ELEMENT_SELECTOR.restaurantAddModal).classList.add("modal--open");
+  }
```
'유저가 개발자 도구로 제어 한다'는 상황을 고려해보지 못한거 같습니다. 🥲

다른 방법들을 찾아보고 적용해 보겠습니다!

jinyoung234 commented on src/components/RestaurantAddForm/RestaurantAddForm.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+  }
+
+  private handleSubmitAddRestaurant(event: Event) {
+    try {
+      event.preventDefault();
+
+      this.addUserInputRestaurantDetail();
+
+      this.handleResetForm();
+
+      this.handleCloseModal();
+
+      this.emit(CUSTOM_EVENT_TYPE.addRestaurant);
+    } catch (error: unknown | Error) {
+      if (error instanceof Error) {
+        alert(error.message);
+      }
+    }
+  }
```
원래 2가지 측면에서 고민해 보았었습니다.

1. Validator를 추가하여 js 단에서 유효성을 검증하는 방법
2. html5의 api를 활용하여 유효성을 검증하는 방법

페어와 고민했을 때 2번을 통해 1번과 같이 가능하며, 1번의 경우 2번과 달리 추가적인 코드 리소스와 복잡성이 커질 것이라고 판단되어 2번의 방식을 채택하게 되었습니다.

다만, 샐리의 코멘트를 듣고 1번의 방식이 필요할 것 같다는 생각이 들었습니다.

또한, `필수 항목들이 입력이 다 되기 전까지는 추가하기 버튼을 비활성화 시키는 것` 또한 `잘못된 값을 추가한다`는 가정이 차단되기 때문에 UX에 많은 도움이 될 거 같습니다.

좋은 방법을 더 찾아본 후 적용해보겠습니다! 감사합니다 :)

jinyoung234 commented on src/components/RestaurantAddForm/RestaurantAddForm.ts:
```diff
@@ -0,0 +1,173 @@
+      }
+    }
+  }
+
```
이것도 제가 고려하지 못했던거 같습니다 🥲🥲🥲🥲🥲

코멘트 반영해서 수정해보겠습니다!!

jinyoung234 commented on src/components/RestaurantAddForm/RestaurantAddForm.ts:
```diff
@@ -0,0 +1,173 @@
+            </div>
+
+            <div class="form-item form-item--required">
+                <label for="name text-caption">이름</label>
+                <input type="text" name="name" id="name" required>
+            </div>
+
+            <div class="form-item form-item--required">
+                <label for="distance text-caption">거리(도보 이동 시간) </label>
+                <select name="distance" id="distance" required>
+                <option value="">선택해 주세요</option>
+                ${createOptionElements(
+                  RestaurantAddForm.DISTANCES_OPTIONS
+                ).replace(
+                  OPTION_ELEMENT_REGEXP,
+                  (_, value) => `<option value=${value}>${value}분 내</option>`
+                )}
+                </select>
+            </div>
+
+            <div class="form-item">
+                <label for="description text-caption">설명</label>
+                <textarea name="description" id="description" cols="30" rows="5"></textarea>
+                <span class="help-text text-caption">메뉴 등 추가 정보를 입력해 주세요.</span>
+            </div>
+
+            <div class="form-item">
+                <label for="url" text-caption">참고 링크</label>
+                <input type="text" name="url" id="url"/>
+                <span class="help-text text-caption">매장 정보를 확인할 수 있는 링크를 입력해 주세요.</span>
+            </div>
+
+            <div class="button-container">
+                <button id="modal-cancel-button" type="button" class="button button--secondary text-caption">취소하기</button>
+                <button type="submit" class="button button--primary text-caption">추가하기</button>
+            </div>
+        </form>
+    `;
+  }
+
```
리액트 처럼 `템플릿 들을 dom에 렌더링 한다.` 한다면 익숙한 형태의 UI를 바라볼 수 있을거 같단 생각에 template을 innerHTML의 형태로 설계했었던거 같습니다.

createElement의 이점을 명확히 파악한건 아니라 조금 더 찾아보고 코멘트 반영해보겠습니다!

jinyoung234 commented on src/components/RestaurantList/RestaurantList.ts:
```diff
@@ -0,0 +1,106 @@
+      (acc: string, restaurantDetail: RestaurantDetail) => {
+        const restaurantItem = new RestaurantItem(restaurantDetail);
+
+        return acc + restaurantItem.getTemplate();
```
`attachShadow`라면 `Shadow DOM`을 의미하시는걸까요?!

제가 아는 Shadow DOM의 장점은 **마크업 구조, 스타일 및 동작을 숨겨서, 페이지의 다른 코드와 충돌하지 않도록 코드를 깨끗하게 유지하기 때문**에, `독립성`**을 보장**해주는데 있다고 생각했습니다.

다만, 스타일의 경우 이미 `css class`로 잘 나눠 놓은 상태기도 하고 만약 충돌이 나더라도 `CSS Modules`로 전환함으로써 해결이 가능하다고 생각합니다.

```jsx
class MyCustomElement extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: 'open' });

    this.shadowRoot.innerHTML = `
      <style>
        .text {
          color: green;
          font-size: 18px;
        }
      </style>
      <div class="text">Shadow DOM 내부의 텍스트</div>
    `;
  }
}
```

저의 경우 **일반 DOM과 shadowRoot를 구분해야 한다는 점**이 사용하면서 불편하다고 느껴졌습니다.

![image](https://github.com/woowacourse/javascript-lunch/assets/87177577/5e513c31-2008-4407-b195-1fcad9963f2f)

개발자 도구에서 shadow root를 열어야 다른 element 들을 확인할 수 있다는 점과, 개발자 도구 내 style을 볼 수 있는 탭이 따로 있음에도 style 태그와 dom element를 동시에 봐야하는게 번거롭다고 느껴져 Shadow DOM을 따로 사용하지 않았던거 같습니다.

샐리가 코멘트 주신 `reduce 돌면서 해주는 것보다 element를 생성`과 `attachShadow를 사용`에 대해 제가 완벽히 이해를 못한거 같은데, 혹시 조금 더 구체적으로 들어볼 수 있을지 여쭤보고 싶습니다. 🥹🙇‍♀️

liswktjs replied:
기존의 코드에서 innerHTML으로 코드를 작성해주고 계셔서 
앞선 innerHTML의 답변과 같은 맥락으로 의문과 함께 innerHTML로 모두 구현하면서 문자열로 관리하고 재사용성을 함께 챙기고 싶다면 
차라리 shadow dom으로 관리하는게 더 낫지 않았을까? 라는 생각이 들어 드린 코멘트였습니다! 
제가 설명이 부족했던 것 같네요

jinyoung234 commented:
@liswktjs 

안녕하세요. 샐리 :) 저번 리뷰 요청 때 컴포넌트의 재사용성을 배제한 것 같이 말씀드린거 같아 우선 죄송하다는 말씀드리고 싶습니다.

- 어플리케이션을 컴포넌트 단위로 모듈화하여 개발
  - UI를 컴포넌트 단위로 생각하고 개발하는 연습
  - 재사용할 수 있는 컴포넌트를 고민해보기

재사용성을 배제했다기보다 lms에 있는 `UI를 컴포넌트 단위로 생각하고 개발하는 연습`를 먼저 목표로 잡고 컴포넌트의 `역할` 관점에서 `하나의 역할`에 대해 오랫동안 고민하다보니 `재 사용할 수 있는 컴포넌트를 고민해보기`에 대한 목표인 `컴포넌트의 재 사용성` 까지 고민 할 시간이 부족했던거 같습니다.

샐리의 리뷰를 토대로 `컴포넌트의 재 사용성` 뿐 아니라`createElement vs innerHTML`, `유효성 검사` 총 3가지 카테고리를 코드 리팩터링을 진행해보았습니다.

# 변경 사항

## 유효성 검사 추가
- 개발자 도구에서 `required` 속성을 제거해도 유효성 검사를 진행할 수 있게 코드를 추가하였습니다.
    - `RestuarantAddForm`(view) → `Resturant`(domain) → `ResturantDetailValidator`(validator)를 통해 유효성 검사를 진행합니다.
        - 유효성 검사를 통과하지 못할 경우 에러를 발생시킵니다.
        - 그 후 에러 메시지를 alert로 먼저 띄웁니다.
        - 그 다음, `ERROR_TARGET_ELEMENTS_DICTIONARY`(에러 메시지에 해당하는 selector를 제공)를 통해 에러 메시지에 해당하는 element에 focus를 제공합니다.

## 재 사용 가능한 컴포넌트 분리
제가 생각한 `재 사용 가능한 컴포넌트`는 아래와 같이 기준을 만들었습니다.

1. 컴포넌트는 하나의 역할을 수행해야 한다.
2. 스타일을 재활용 할 수 있어야 한다.

이 2가지를 최우선 목표로 설정 후 `modal`과 `dropdown`, `formitem`에 대한 공통 컴포넌트를 만들었습니다.

각 컴포넌트는 아래와 같은 특징을 가집니다.

**modal**
- 열고 닫을 수 있다.
- backdrop이 내장되어 잇다.
- children을 받아 확장이 가능하다.
- 공통된 style을 가지며 style 확장이 가능하다.

**dropdown**
- option을 선택할 수 있다.
- 공통된 style을 가지며 style 확장이 가능하다.
- eventType과 target을 받아 custom event를 dispatch 할 수 있다.

**formitem**
- 공통된 style을 가지며 style 확장이 가능하다.

이 컴포넌트를 base로 확장할 수 있게 코드를 구성했습니다. 

## 개발자 도구에서 modal 제어하지 못하도록 변경

modal 또한 개발자 도구에서 제어가 가능하기 때문에 `dialog` 태그를 통해 모달을 열고 닫는 형태로 리팩터링 해보았습니다.

## 애플리케이션 내 type-safe 요소 추가

- 타입스크립트가 제공하는 `is`를 활용하여 타입 좁히기를 해보았습니다.
- `const assertion`을 추가하여 자동 완성 및 read만 가능하도록 객체 타입을 변경해보았습니다. 

# 궁금한 점

## innerHTML vs createElement & appendChild

 샐리가 코멘트 주신 부분에 대해 최선을 다해 고민을 해보았습니다.

 우선, 코멘트 주신 대로 `createElement`와 `appendChild`를 활용하여 컴포넌트를 수정해보았습니다.

<details>
<summary>innerHTML</summary>

```jsx
import "./LunchApp.css";

import BaseComponent from "../BaseComponent/BaseComponent";

class LunchApp extends BaseComponent {
  protected render() {
    this.innerHTML = `
          <global-navigation-bar></global-navigation-bar>
          <main>
            <section class="restaurant-filter-container">
              <category-dropdown></category-dropdown>            
              <sort-dropdown></sort-dropdown>
            </section>
            <restaurant-list></restaurant-list>
          </main>
          <restaurant-add-modal class="modal"></restaurant-add-modal>
        `;
  }
}

customElements.define("lunch-app", LunchApp);
```  

</details>

<details>
<summary>createElement & appendChild</summary>

<br/>

**초기 구조**
      
```jsx
import "./LunchApp.css";

import BaseComponent from "../BaseComponent/BaseComponent";

import { createElements } from "../../utils/createElements";

class LunchApp extends BaseComponent {
  render() {
    const fragmentNode = document.createDocumentFragment();

    const globalNavBar = document.createElement("global-navigation-bar");
    fragmentNode.appendChild(globalNavBar);

    const mainSection = document.createElement("main");

    const filterContainer = document.createElement("section");
    filterContainer.className = "restaurant-filter-container";

    const categoryDropdown = document.createElement("category-dropdown");
    filterContainer.appendChild(categoryDropdown);

    const sortDropdown = document.createElement("sort-dropdown");
    filterContainer.appendChild(sortDropdown);

    mainSection.appendChild(filterContainer);

    const restaurantList = document.createElement("restaurant-list");
    mainSection.appendChild(restaurantList);

    fragmentNode.appendChild(mainSection);

    const addModal = document.createElement("restaurant-add-modal");
    addModal.className = "modal";
    fragmentNode.appendChild(addModal);

    this.appendChild(fragmentNode);
  }
}

customElements.define("lunch-app", LunchApp);
```

 저의 경우 컴포넌트를 설계할 때 보기 쉬운 형태인 선언적인 UI를 만들고 싶었기 때문에 지금의 형태는 너무 절차적으로 UI가 만들어진다는 생각이 들었습니다.
    
 제 생각이 잘못되었나 싶어 다른 크루들의 의견을 물어봐도 구조적으론 `innerHTML`의 형태가 더 보기 좋다는 의견이 지배적 이었습니다.
    
 그래서 `react`의 `createElement`에 영감을 받아 다음과 같이 구조를 수정해보았습니다.
    
**React의 createElement 모방**
        
```jsx
class LunchApp extends BaseComponent {
  render() {
    this.appendChild(
      createElements({
        tag: "fragment",
        children: [
          "global-navigation-bar",
          {
            tag: "main",
            children: [
              {
                tag: "section",
                attrs: { className: "restaurant-filter-container" },
                children: ["category-dropdown", "sort-dropdown"],
              },
              "restaurant-list",
            ],
          },
          {
            tag: "restaurant-add-modal",
            attrs: { className: "modal" },
          },
        ],
      })
    );
  }
}
```
  
 이전과 달리 조금 구조적으로 보기 좋아졌지만, 이전과 비교하면 XSS 공격에 덜 취약 하다는 것외에 여전히 구조적으로 보기 좋다는 생각이 들지는 않았으며, 어떤 장점이 있는지 확인하기 어려웠습니다. 🥲
    
</details>

 그래서 여러 자바스크립트 서적을 참고해보게 되었고 제가 생각한 `innerHTML`과 `createElement & appendChild(DOM을 직접 조작)`의 차이점은 아래와 같다고 생각했습니다.

<h3><strong>innerHTML</strong></h3>

장점 | 단점
-- | --
DOM 조작 방식에 비해 빠르고 간편하다. | XSS공격에 취약점이 있기 때문에 사용자로 부터 입력받은 콘텐츠(untrusted data: 댓글, 사용자 이름 등)를 추가할 때 주의하여야 한다.
간편하게 문자열로 정의한 여러 요소를 DOM에 추가할 수 있다. | 해당 요소의 내용을 덮어 쓴다. 즉, HTML을 다시 파싱한다. 이것은 비효율적이다. (reflow, repaint)

<h3><strong>DOM 조작 방식</strong></h3>

장점 | 단점
-- | --
특정 노드 한 개(노드, 텍스트, 데이터 등)를 DOM에 추가할 때 적합하다. | innerHTML보다 느리고 더 많은 코드가 필요하다.

innerHTML의 단점은 크게 2가지 라고 생각했습니다.

1. reflow, repaint 과정이 때에 따라 DOM 조작 방식에 비해 더 많이 발생할 수 있다.
2. XSS 공격에 취약하다.

그래서 개발자 도구의 성능 탭을 활용하여 두 API의 성능 차이를 비교해보았습니다. (0 ~ 1.3초 까지의 측정 결과)

<details>
<summary>innerHTML</summary>

<details>
<summary>요약</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/317800a3-a551-4cb4-a9b3-97b433e559fc"/>
</details>

<details>
<summary>레이아웃</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/946a0eb8-8036-46b5-ab43-41614badc63e"/>
</details>

<details>
<summary>페인팅</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/3459a1f1-f810-4977-948f-530c55c11eac"/>
</details>

</details>

<details>
<summary>createElement & appendChild</summary>

<details>
<summary>요약</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/2cb56ec2-74fa-4ee1-bf67-d8d27e75fd6d"/>
</details>

<details>
<summary>레이아웃</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/347cae7d-4c6a-46cf-94f0-f648eea9fabb"/>
</details>

<details>
<summary>페인팅</summary>
<img src="https://github.com/woowacourse/javascript-lunch/assets/87177577/f5ded1ec-2ba2-47af-b9b4-c003b72c2283"/>
</details>

</details>

 1번의 경우 `innerHTML`과 성능적으로 차이가 없다 싶을 만큼 미미했습니다.

 그렇다면 2번의 이유에서 바꿔볼 수 있을거 같은데 2번의 관점에서도 고민해보았을 때, [DOMPurify](https://github.com/cure53/DOMPurify) 라이브러리를 통해 대체 가능 하다고 생각했습니다.

 즉, 아직까지 `createElement`와 `appendChild`의 장점에 대해 명확해지지 않은 상태인데, 샐리가 생각했을 때의 `createElement`의 장점과 효용에 대해 조금 더 이야기를 들어보고 싶습니다. 🥹

## 웹 컴포넌트
### 도입 이유
제가 웹 컴포넌트를 도입했던 이유는 처음 바닐라 js로 웹 애플리케이션을 만들 때 MVC 패턴을 적용하며 느꼈던 `불편함` 때문이었습니다.

구현 순서는 아래와 같았습니다.

1. index.html에 모든 element를 모두 넣어넣는다.
2. MVC 패턴으로 구현을 진행하며 View 에서는 DOM을 취득하여 조작하는 레이어를 분리한다.
3. 컴포넌트 렌더링 및 이벤트 등록을 진행한다.

이렇게 구현했을 때 문제점은 아래와 같았습니다.

1. index.html에 있는 모든 template 들이 어떤 영역의 template인지 파악하기 어렵다.
2. controller의 책임이 너무 많아진다. (view에 데이터 전달, domain - controller 간 비즈니스 로직 주고 받기 등)
3. 렌더링 방법 및 이벤트 등록 방법이 너무 절차적이다.

그래서 작은 단위로 나눠 각각의 컴포넌트가 스스로 이벤트 핸들러와 render할 template 들을 관리하기 위한 방법 들을 고민했고 Web Component의 `생명 주기 함수`에 대해 주목하였습니다.

Web Component를 통해 느끼고자 했던 이점은 아래와 같았습니다.

1. UI 내부에서 template 들을 관리함으로써, 컴포넌트가 어떤 구조인지 명확히 파악한다.
2. 라이프사이클 메서드를 통해 render 및 이벤트 등록을 컴포넌트가 mount 되었을 때 선언적으로 진행할 수 있다.
3. 컴포넌트의 관심사를 명확히 나눌 수 있다.

그 결과 Web Component를 사용하게 되었습니다.

하지만 사용하면서 아래와 같은 불편함을 느꼈습니다.

### 불편한 컴포넌트 확장

<details>
<summary>Common Modal</summary>

```jsx
import "./CommonModal.css";

import BaseComponent from "../../BaseComponent/BaseComponent";

class CommonModal extends BaseComponent {
  constructor() {
    super();

    this.attachShadow({ mode: "open" });
  }

  protected render() {
    if (!this.shadowRoot) return;

    this.innerHTML = `
        <style>      
          .modal-backdrop {
            position: fixed;
            top: 0;
            right: 0;
            bottom: 0;
            left: 0;
          
            background: rgba(0, 0, 0, 0.35);
          }
          
          .modal-container {
            position: fixed;
            bottom: 0;
            width: 100%;
            max-height: 620px;
          
            padding: 24px 16px;
            overflow: auto;
          
            border-radius: 8px 8px 0px 0px;
            background: var(--grey-100);
          }
          
          /* Chrome, Safari */
          .modal-container::-webkit-scrollbar {
            display: none;
          }
          
          /* Firefox */
          .modal-container {
            scrollbar-width: none;
          }
          
          /* IE & Edge */
          .modal-container {
            -ms-overflow-style: none;
          }    
        </style>
        
        <div class="modal-backdrop"></div>
        <div class="modal-container">
          <slot name="modal-content"></slot>
        </div>
      `;
  }
}

customElements.define("common-modal", CommonModal);
```
</details>

<details>
<summary>RestaurantAddForm</summary>

```jsx
class RestaurantAddForm extends BaseComponent {
  // ...

  constructor() {
    super();

    this.attachShadow({ mode: "open" });
  }

  protected render(): void {
    if (!this.shadowRoot) return;

    const menuCategoryWithoutAllOptions =
      Object.values(MENU_CATEGORIES).slice(1);

    this.innerHTML = `
        <style>
          .form-item {
            display: flex;
            flex-direction: column;
          
            margin-bottom: 24px;
          }
          
          .form-item label {
            color: var(--grey-400);
            font-size: 14px;
          }
          
          .form-item--required label::after {
            padding-left: 4px;
          
            color: var(--primary-color);
            content: "*";
          }
          
          .form-item .help-text {
            color: var(--grey-300);
          }
          
          .form-item input,
          .form-item textarea,
          .form-item select {
            padding: 8px;
            margin: 6px 0;
          
            border: 1px solid var(--grey-200);
            border-radius: 8px;
          
            font-size: 16px;
          }
          
          .form-item textarea {
            resize: none;
          }
          
          .form-item select {
            width: 100%;
            height: 44px;
          
            padding: 8px;
          
            border: 1px solid var(--grey-200);
            border-radius: 8px;
          
            color: var(--grey-300);
          }
          
          input[name="name"],
          input[name="link"] {
            height: 44px;
          }
          
          .button-container {
            display: flex;
          }
          
          .button {
            width: 100%;
            height: 44px;
          
            margin-right: 16px;
          
            border: none;
            border-radius: 8px;
          
            font-weight: 600;
            cursor: pointer;
          }
          
          .button:last-child {
            margin-right: 0;
          }
          
          .button--secondary {
            border: 1px solid var(--grey-300);
            background: transparent;
          
            color: var(--grey-300);
          }
          
          .button--primary {
            background: var(--primary-color);
          
            color: var(--grey-100);
          }
        </style>
        <form id="restaurant-add-form">
            <div class="form-item form-item--required">
                <label for="category text-caption">카테고리</label>
                <common-dropdown id="category" options="${menuCategoryWithoutAllOptions}" title="선택해 주세요" /></common-dropdown>            
            </div>

            <div class="form-item form-item--required">
                <label for="name text-caption">이름</label>
                <input type="text" name="name" id="name" required>
            </div>

            <div class="form-item form-item--required">
                <label for="distance text-caption">거리(도보 이동 시간) </label>
                <common-dropdown addOptionText="분 내" id="distance" options="${RestaurantAddForm.DISTANCES_OPTIONS}" title="선택해 주세요" /></common-dropdown>            
            </div>

            <div class="form-item">
                <label for="description text-caption">설명</label>
                <textarea name="description" id="description" cols="30" rows="5"></textarea>
                <span class="help-text text-caption">메뉴 등 추가 정보를 입력해 주세요.</span>
            </div>

            <div class="form-item">
                <label for="url" text-caption">참고 링크</label>
                <input type="text" name="url" id="url"/>
                <span class="help-text text-caption">매장 정보를 확인할 수 있는 링크를 입력해 주세요.</span>
            </div>

            <div class="button-container">
                <button id="modal-cancel-button" type="button" class="button button--secondary text-caption">취소하기</button>
                <button type="submit" class="button button--primary text-caption">추가하기</button>
            </div>
        </form>
    `;
    // ...
  }
```
</details>

<details>
<summary>LunchApp</summary>

```jsx
// lunchApp.ts
<common-modal
   <div slot="modal-content">
       <h2 class='modal-title text-title'>새로운 음식점</h2>
       <restaurant-add-form></restaurant-add-form>
   </div>
</common-modal>
```
</details>

`slot`을 사용할 경우 공통 컴포넌트와 공통 컴포넌트를 사용하는 컴포넌트는 필연적으로 `shadow dom`을 적용해야했습니다.

이렇다보니 render 내부의 template 들이 불필요하게 증가되었으며, shadowRoot 내 타입 좁히기도 필연적이게 되어 코드가 매우 복잡해진다는 단점이 있었습니다.

또한 `shadow root`를 사용하다보니 `querySelector` 같은 dom 취득 메서드를 사용할 때 비정상적으로 작동하여 디버깅에도 어려움을 겪었습니다.

그래서 결론적으론 attribute를 통해 공통 컴포넌트에 데이터를 내려주도록 하였습니다.

하지만 이 또한, 많은 문제점이 있다고 느껴졌습니다.

### 수 많은 custom elements로 인한 개발자 도구 사용의 어려움

<img width="799" alt="image" src="https://github.com/woowacourse/javascript-lunch/assets/87177577/9be88637-d1fe-4d75-8748-4b28f2a9be00">

컴포넌트를 확장하고 또 확장하다보니 불필요한 html element가 과도하게 많아진다는 느낌을 받게 되었습니다.

위 사진과 같이 select 태그를 살펴보기 위해 깊은 depth를 내리며 찾아봐야한다는 불편함을 느끼게 되었습니다.

## 최종적인 생각

이번 미션에서 이 기술을 도입하며 장/단점을 정말 명확하게 느꼈다고 생각되는데, 구조적인 코드를 선호하는 제 입장에선 그래도 아직까지는 단점보다는 장점이 더 뚜렷하다는 생각이 들었습니다.

하지만 제 생각을 확고히 하는 것 보다는 다른 분들의 의견을 통해 좀 더 인사이트를 받아보고 싶은데 샐리가 느꼈던 웹 컴포넌트의 장점과 단점은 어떤 것들이 있었는지 궁금합니다. (답변이 어렵다면 안해주셔도 괜찮습니다!)
